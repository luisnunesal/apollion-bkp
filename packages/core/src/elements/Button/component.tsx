import React, { forwardRef } from 'react';

import { ButtonInterface } from './interface';
import { ButtonStyle } from './style';

import { Content } from '../../containers/Content';
import { BaseText } from '../Typography';

export const ButtonComponent = forwardRef<HTMLButtonElement, ButtonInterface>(
  ({ children, text, isLoading, icon, iconPosition, ...props }, ref) => (
    <ButtonStyle {...props} as="button" ref={ref}>
      <Content loading={isLoading} {...(iconPosition === 'right' ? { closer: icon } : { opener: icon })}>
        <BaseText fontWeight="bold" lineHeight="tight" fontFamily="main">
          {children || text}
        </BaseText>
      </Content>
    </ButtonStyle>
  ),
);

ButtonComponent.defaultProps = {
  color: 'primary',
  size: 'medium',
  variant: 'contained',
  icon: null,
  iconPosition: 'left',
  fullWidth: false,
};
