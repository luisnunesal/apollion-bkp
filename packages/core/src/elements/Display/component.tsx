import React, { forwardRef } from 'react';

import { DisplayComponentInterface } from './interface';
import { DisplayStyle } from './style';

export const Display = forwardRef<HTMLHeadingElement, DisplayComponentInterface>(
  ({ variant = 'h2', children, text, ...props }, ref) => {
    const content = children ?? text;

    return (
      <DisplayStyle as={variant} variant={variant} ref={ref} {...props}>
        {content}
      </DisplayStyle>
    );
  },
);
