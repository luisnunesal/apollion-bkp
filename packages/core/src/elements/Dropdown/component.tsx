import React, { useEffect, useState } from 'react';
import Tippy, { TippyProps } from '@tippyjs/react/headless';

import { DropdownPosition, DropdownProps, RenderProps } from './interface';

import { Paper } from '../../containers/Paper';
import { useToggle } from '../../hooks';
import { isFunction } from '../../utils';
import { List } from '../List';

const dropdownPosition: Record<DropdownPosition, TippyProps['placement']> = {
  left: 'bottom-start',
  right: 'bottom-end',
  center: 'bottom',
};

export const DropdownComponent = ({ children, options, content, hoverable, position, width, full }: DropdownProps) => {
  if (!options && !content) {
    throw new Error('Dropdown deve receber "content" ou "options" props');
  }

  const [triggerNode, getTriggerNode] = useState<HTMLElement>();
  const [triggerWidth, setTriggerWidth] = useState<number | undefined>();

  const { active: visible, enable: openDropdown, disable: closeDropdown, toggle } = useToggle(false);

  useEffect(() => {
    if (full && triggerNode) {
      const rect = triggerNode.getBoundingClientRect();
      setTriggerWidth(rect.width);
    }
  }, [triggerNode]);

  const renderChildren = () => {
    const itIsFunction = isFunction(children);

    const element = isFunction(children)
      ? (children as RenderProps)({ isOpen: visible, open: openDropdown, toggle })
      : children;

    const childrenProps = {
      ref: getTriggerNode,
      ...(!itIsFunction && {
        onClick: toggle,
      }),
    };

    return React.cloneElement(element as React.ReactElement, childrenProps);
  };

  const renderDropdown: TippyProps['render'] = (attrs) => {
    let contentElement = content;

    if (Array.isArray(options)) {
      contentElement = <List gap="zero" color="primary.action" content={options} />;
    }

    if (isFunction(content)) {
      contentElement = content({ close: closeDropdown });
    }

    return (
      <Paper p="xs" borderRadius="xs" deep={2} width={width ?? triggerWidth} {...attrs}>
        {contentElement}
      </Paper>
    );
  };

  const customProps: Partial<TippyProps> = {
    ...(!hoverable && {
      onClickOutside: closeDropdown,
      visible,
    }),
  };

  return (
    <>
      {renderChildren()}
      <Tippy
        {...customProps}
        interactive
        render={renderDropdown}
        reference={triggerNode}
        placement={dropdownPosition[position]}
      />
    </>
  );
};

DropdownComponent.defaultProps = {
  position: 'left',
  hoverable: false,
};
