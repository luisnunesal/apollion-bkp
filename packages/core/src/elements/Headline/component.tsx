import React, { forwardRef } from 'react';

import { HeadlineComponentInterface } from './interface';
import { HeadingStyle } from './style';

export const Headline = forwardRef<HTMLHeadingElement, HeadlineComponentInterface>(
  ({ variant = 'h2', children, text, ...props }, ref) => {
    const content = children ?? text;

    return (
      <HeadingStyle as={variant} ref={ref} variant={variant} {...props}>
        {content}
      </HeadingStyle>
    );
  },
);
