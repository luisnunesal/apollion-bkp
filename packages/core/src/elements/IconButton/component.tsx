import React, { forwardRef } from 'react';

import { IconButtonInterface } from './interface';
import { IconButtonStyle } from './style';

import { Spinner } from '../Spinner';

export const IconButtonComponent = forwardRef<HTMLButtonElement, IconButtonInterface>((props, ref) => {
  const { isLoading, loadingComponent, icon } = props;

  return (
    <IconButtonStyle {...props} as="button" ref={ref}>
      {isLoading ? loadingComponent ?? <Spinner data-testid="icon-spinner" /> : icon}
    </IconButtonStyle>
  );
});

IconButtonComponent.defaultProps = {
  color: 'primary',
  size: 'medium',
  variant: 'contained',
  icon: null,
};
