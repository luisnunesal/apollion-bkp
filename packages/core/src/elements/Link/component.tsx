import React, { forwardRef } from 'react';

import { LinkButtonInterface, LinkInterface } from './interface';
import { LinkButtonStyle, LinkStyle } from './style';

import { BaseText } from '../Typography';

const getRoute = ({ href, to }: LinkInterface): LinkInterface['to'] => (href ? { pathname: href } : to);

const LinkButtonComponent = forwardRef<HTMLAnchorElement, LinkButtonInterface>(({ text, children, ...props }, ref) => {
  const toRoute = getRoute(props);

  return (
    <LinkButtonStyle {...props} to={toRoute} ref={ref}>
      <BaseText fontWeight="bold" lineHeight="tight">
        {children || text}
      </BaseText>
    </LinkButtonStyle>
  );
});

LinkButtonComponent.defaultProps = {
  variant: 'contained',
  color: 'primary',
};

export const LinkComponent = forwardRef<HTMLAnchorElement, LinkInterface>(({ text, children, ...props }, ref) => {
  const toRoute = getRoute(props);

  return (
    <LinkStyle {...props} to={toRoute} ref={ref}>
      {children || text}
    </LinkStyle>
  );
}) as React.ForwardRefExoticComponent<LinkInterface & React.RefAttributes<HTMLAnchorElement>> & {
  Button: typeof LinkButtonComponent;
};

LinkComponent.Button = LinkButtonComponent;

LinkComponent.defaultProps = {
  color: 'primary',
};
