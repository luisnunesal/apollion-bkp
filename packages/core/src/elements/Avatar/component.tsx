import React from 'react';

import { getIconSize } from './helpers';
import { AvatarProps } from './interface';
import { AvatarContainer } from './style';

import { Icon } from '../Icon';
import { Image } from '../Image';
import { Typography } from '../Typography';

function getContent({ imgSrc, icon, size, label }: Partial<AvatarProps>) {
  if (imgSrc) {
    return <Image width="100%" height="100%" src={imgSrc} data-testid="avatar-image" />;
  }

  if (icon) {
    return <Icon name={icon} size={getIconSize(size)} data-testid="avatar-icon" />;
  }

  return <Typography textTransform="uppercase">{label}</Typography>;
}

export const Avatar: React.FC<AvatarProps> = ({ size, icon, imgSrc, label, ...props }) => (
  <AvatarContainer
    alignItems="center"
    justifyContent="center"
    overflow="hidden"
    borderRadius="circular"
    bgColor="grayscale.5"
    color="primary.base"
    label={label}
    size={size}
    {...props}
  >
    {getContent({ size, icon, imgSrc, label })}
  </AvatarContainer>
);
