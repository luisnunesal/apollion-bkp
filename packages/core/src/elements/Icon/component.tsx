import React, { forwardRef, useEffect, useState } from 'react';

import { getIconFromUrl } from './helper';
import { IconProps, IconVariantProps } from './interface';
import { IconStyle } from './style';

import { icons } from '../../icons';
import { Image } from '../Image';
import { SvgComponent } from '../Svg/component';

const createIcon = (element: React.ReactNode, props: any) => React.cloneElement(element as React.ReactElement, props);

const IconRegular = forwardRef<HTMLDivElement, IconVariantProps>(
  ({ icon, name, iconProps, svgProps, children, ...props }, ref) => {
    const getIcon = () => {
      let iconComponent = null;

      if (children) {
        iconComponent = createIcon(children as React.ReactElement, svgProps);
      } else if (name) {
        iconComponent = createIcon(icons[name], svgProps);
      } else if (icon) {
        iconComponent = createIcon(icon as React.ReactElement, svgProps);
      } else if (iconProps) {
        iconComponent = SvgComponent(svgProps);
      }

      return iconComponent;
    };

    return (
      <IconStyle {...props} ref={ref}>
        {getIcon()}
      </IconStyle>
    );
  },
);

const IconUrl = forwardRef<HTMLDivElement, IconVariantProps>(({ title, src, svgProps, ...props }, ref) => {
  const [iconElement, setIcon] = useState<React.ReactNode>(() => <Image height="1em" width="1em" alt={title} />);

  useEffect(() => {
    getIconFromUrl(src, (element) => {
      setIcon(createIcon(element, svgProps));
    });
  }, [src]);

  return (
    <IconStyle {...props} ref={ref}>
      {iconElement}
    </IconStyle>
  );
});

export const IconComponent: React.FC<IconProps> = forwardRef<HTMLDivElement, IconProps>(({ title, ...props }, ref) => {
  const svgProps = {
    ...(props.iconProps || {}),
    title,
    stroke: 'currentColor',
    fill: 'currentColor',
    strokeWidth: '0',
    height: '1em',
    width: '1em',
  };

  if (props.src) {
    return <IconUrl {...props} svgProps={svgProps} ref={ref} />;
  }

  return <IconRegular {...props} svgProps={svgProps} ref={ref} />;
});
