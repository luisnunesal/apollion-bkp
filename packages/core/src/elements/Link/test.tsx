import React from 'react';
import { MemoryRouter as Router } from 'react-router-dom';

import { LinkComponent as Link } from './component';

import { render, screen } from '../../__tests__/setup';

// MOCKS
const mockHistoryPush = jest.fn();
jest.mock('react-router-dom', () => ({
  // @ts-ignore
  ...jest.requireActual('react-router-dom'),
  useHistory: () => ({
    push: mockHistoryPush,
  }),
}));

describe('Link', () => {
  // ---- SNAPSHOT ---- //
  it('Should match snapshot', () => {
    render(
      <Router>
        <Link href="#!" text="Hello World" />
      </Router>,
    );
    expect(screen).toMatchSnapshot();
  });

  // ---- CONTENT ---- //
  it('Should render with text="Hello World"', () => {
    render(
      <Router>
        <Link href="#!" text="Hello World" />
      </Router>,
    );
    expect(screen.getByText('Hello World')).toBeDefined();
  });

  it('Should render with children equals "Hello World"', () => {
    render(
      <Router>
        <Link href="#!">Hello World</Link>
      </Router>,
    );
    expect(screen.getByText('Hello World')).toBeDefined();
  });

  it('Should prioritize children instead of text', () => {
    render(
      <Router>
        <Link href="#!" text="Text">
          Children
        </Link>
      </Router>,
    );
    expect(screen.getByText('Children')).toBeDefined();
  });
});
