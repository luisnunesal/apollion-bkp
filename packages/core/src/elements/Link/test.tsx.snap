// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link Should match snapshot 1`] = `
<a
  class=""
  href="#!"
>
  Hello World
</a>
`;
